File: pege/components/fetcher.py
from .screen import Screen
from .mmu import MMU



class Fetcher:

    READ_TILE = 0
    READ_DATA_0 = 1
    READ_DATA_1 = 2
    PUSH_PIXELS = 3

    LY_REGISTER = 0xFF44
    LCDC_REGISTER = 0xFF40



    def __init__(self, mmu, ppu):
        self._mmu = mmu
        self.state = Fetcher.READ_TILE
        self.ppu = ppu
        self.offset = MMU.BACKGROUND_MAP_START
        self.tilemap_ram = MMU.TILERAM_START
        self.tile_address = None
        self.column = 0
        self.ticks = 0
    def update_row(self, scy):
        if scy % 8 == 0:
                self.offset += 32

    def update_column(self, scx):
        self.column = scx

    def _decode(self, fb, sb):
        result = []
        for x in range(7, -1, -1):
            msb = fb >> x & 0x01
            lsb = sb >> x & 0x01

            if msb and lsb:
                result.append(Screen.DARK_GREEN)
            elif not msb and not lsb:
                result.append(Screen.LIGHTEST_GREEN)
            elif msb and not lsb:
                result.append(Screen.DARKEST_GREEN)
            elif lsb and not msb:
                result.append(Screen.LIGHT_GREEN)
        return result


    def _read_tile(self):
        LY = self._getLY()
        #LY = self.ppu.get_LY()
        current_offset = self.offset + self.column
        tile_id = self._mmu.read(current_offset)
        self.tile_address = self.tilemap_ram + (tile_id * 16)

    def _getLY(self):
        return self._mmu.read(Fetcher.LY_REGISTER)

    def _read_data_0(self):
        address = self.tile_address
        LY = self._getLY()
        offset = (LY % 8)
        offset += offset
        tile_row = address + offset
        b = tile_row

        self.data_0 = self._mmu.read(b)

    def _read_data_1(self):
        address = self.tile_address
        b = address + 1
        self.data_1 = self._mmu.read(b)


    def step(self):
            
        self.ticks += 1
        if self.ticks < 2:
            return
        self.ticks = 0

        if self.state == Fetcher.READ_TILE:
            self._read_tile()
            self.state = Fetcher.READ_DATA_0

        elif self.state == Fetcher.READ_DATA_0:
            self._read_data_0()
            self.state = Fetcher.READ_DATA_1
        elif self.state == Fetcher.READ_DATA_1:
            self._read_data_1()
            self.state = Fetcher.PUSH_PIXELS

    def reset_fetcher(self):
        self.state = Fetcher.READ_TILE
        self.offset = MMU.BACKGROUND_MAP_START
        self.tilemap_ram = MMU.TILERAM_START
        self.tile_address = None
        self.column = 0
        self.ticks = 0


    def reset_column(self):
        self.column = 0

    def fetch():
        return self._decode(self.data_0, self.data_1)

File: ekkel.gameboy/src/main/java/ekkel/gameboy/ppu/Fetcher.java
package ekkel.gameboy.ppu;

import core.cpu.opcodes.exceptions.NotImplementedException;
import core.mmu.*;
import ekkel.gameboy.cpu.IOMap;
import gui.GameBoyDMGPalette;

import java.util.ArrayList;
import java.util.List;

public class Fetcher {


    FetcherState state;
    private MMU mmuImpl;

    private int offset;
    private int tilemapRam;
    private int tileAddress;
    private int column;
    private int ticks;

    private int tileMapOffset = 0;
    private int bgMapOffset = 0;
    private MemoryValue data0;
    private MemoryValue data1;

    public Fetcher(MMU mmuImpl) {
        this.state = FetcherState.READ_TILE;
        this.mmuImpl = mmuImpl;
        this.column = 0;
        this.ticks = 0;
    }

    private Computable getLY() throws NotImplementedException, IllegalAccessException {
        return this.mmuImpl.read(MemoryAddress.fromValue(IOMap.LY));
    }

    private void readTile() throws NotImplementedException, IllegalAccessException {
        int currentOffset = this.offset + this.column;
        int tileId = this.mmuImpl.read(MemoryAddress.fromValue(currentOffset)).getValue();

        this.tileAddress = tilemapRam + (tileId * 16);
    }

    private void readData0() throws NotImplementedException, IllegalAccessException {
        int tileAddress = this.tileAddress;
        Computable LY = getLY();
        int o = (LY.getValue() % 8);
        o += o;
        int tileRow = tileAddress + o;
        this.data0 = this.mmuImpl.read(MemoryAddress.fromValue(tileRow));
    }

    private void readData1() throws NotImplementedException, IllegalAccessException {
        int address = tileAddress + 1;
        this.data1 = this.mmuImpl.read(MemoryAddress.fromValue(address));
    }

    public Computable getData0() {
        return this.data0;
    }

    public Computable getData1() {
        return this.data1;
    }

    public List<Integer> decode(Computable fb, Computable sb) {
        List<Integer> result = new ArrayList<>();
        for (int x = 7; x > -1; x--) {
            int msb = fb.getValue() >> x & 0x01;
            int lsb = sb.getValue() >> x & 0x01;

            if (msb == 0x01 && lsb == 0x01) {
                result.add(GameBoyDMGPalette.DARK_GREEN);
            } else if (msb != 0x01 && lsb != 0x01) {
                result.add(GameBoyDMGPalette.LIGHTEST_GREEN);
            } else if (lsb == 0x01) {
                result.add(GameBoyDMGPalette.LIGHT_GREEN);

            } else {
                result.add(GameBoyDMGPalette.DARKEST_GREEN);
            }
        }

        return result;
    }

    public void reset() {
        this.state = FetcherState.READ_TILE;
        this.tilemapRam = tileMapOffset;
        this.offset = bgMapOffset;
        this.tileAddress = 0;
        this.column = 0;
        this.ticks = 0;
    }

    public void setOffset(int offset) {
        this.offset = offset;
        this.bgMapOffset = offset;
    }

    public void setTilemapRam(int tmr) {
        this.tilemapRam = tmr;
        this.tileMapOffset = tmr;
    }


    public void updateColumn(int scx) {
        this.column = scx;
    }

    public void updateRow(int scy) {
        if (scy % 8 == 0 && scy != 0) {
            this.offset += 32;
        }
    }

    public void tick() throws NotImplementedException, IllegalAccessException {
        ticks += 1;
        if (this.ticks < 2) {
            return;
        }
        this.ticks = 0;
        switch (state) {
            case READ_TILE:
                readTile();
                this.state = FetcherState.READ_DATA_0;
                break;
            case READ_DATA_0:
                readData0();
                this.state = FetcherState.READ_DATA_1;
                break;
            case READ_DATA_1:
                readData1();
                this.state = FetcherState.PUSH_PIXELS;
                break;
        }
    }


}
